src/transformers/models/mistral/modeling_tf_mistral.py

```diff
@@ -0,0 +1,1068 @@
+        """
+        TFMistralRMSNorm is equivalent to T5LayerNorm
+        """
+        super().__init__(**kwargs)
+        self.hidden_size = hidden_size
+        self.variance_epsilon = eps
+
+    def build(self, input_shape=None):
+        self.weight = self.add_weight(
+            name="weight",
+            shape=self.hidden_size,
+            initializer="ones",
+        )
+        if self.built:
+            return
+        self.built = True
+
+    def call(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        hidden_states = tf.cast(hidden_states, tf.float32)
+        variance = tf.reduce_mean(tf.square(hidden_states), axis=-1, keepdims=True)
+        hidden_states = tf.divide(hidden_states, tf.sqrt(variance + self.variance_epsilon))
+        return self.weight * tf.cast(hidden_states, input_dtype)
+
+
+# Verification: https://colab.research.google.com/gist/ariG23498/f8d8131b795a131b93d99e70ee93c192/scratchpad.ipynb
+class TFMistralRotaryEmbedding(keras.layers.Layer):
+    def __init__(self, dim, max_position_embeddings=2048, base=10000, **kwargs):
+        super().__init__(**kwargs)
+
+        self.dim = dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        self.inv_freq = 1.0 / (self.base ** (tf.range(start=0, limit=self.dim, delta=2, dtype=tf.float32) / self.dim))
+        self._set_cos_sin_cache(seq_len=max_position_embeddings, dtype=tf.float32)
+
+    def _set_cos_sin_cache(self, seq_len, dtype):
+        self.max_seq_len_cached = seq_len
+        t = tf.cast(tf.range(self.max_seq_len_cached, dtype=tf.int64), self.inv_freq.dtype)
+
+        # freqs = tf.tensordot(t, self.inv_freq, axes=0)
+        freqs = tf.einsum("i,j->ij", t, self.inv_freq)
+        # Different from paper, but it uses a different permutation in order to obtain the same calculation
+        emb = tf.concat([freqs, freqs], axis=-1)
+
+        self.cos_cached = tf.cast(tf.cos(emb), dtype)
+        self.sin_cached = tf.cast(tf.sin(emb), dtype)
+
+    def call(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, dtype=x.dtype)
+
+        cos_values = tf.gather(self.cos_cached, tf.range(seq_len))
+        cos_values = tf.cast(cos_values, dtype=x.dtype)
+        sin_values = tf.gather(self.sin_cached, tf.range(seq_len))
+        sin_values = tf.cast(sin_values, dtype=x.dtype)
+        return (cos_values, sin_values)
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    end_index = shape_list(x)[-1]
+    mid_index = end_index // 2
+    indices1 = tf.range(mid_index)  # Indices for the first half
+    indices2 = tf.range(mid_index, end_index)  # Indices for the second half
+    x1 = tf.gather(x, indices1, axis=-1)
+    x2 = tf.gather(x, indices2, axis=-1)
+    return tf.concat([-x2, x1], axis=-1)
```

Comment:
There might be a simpler way to do this - check the `tf_gemma` code! You can steal that code with a `Copied from` if it works.

Reply:
Using the following code from `tf_gemma`
```py
def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., : tf.shape(x)[-1] // 2]
    x2 = x[..., tf.shape(x)[-1] // 2 :]
    return tf.concat([-x2, x1], axis=-1)
```
results in the `TFMistralModelTest::test_embeddings_out_of_bounds_raise_exception` test to fail. I think it is supposedly due to the slicing operation. The `gather` operation raises an excpetion which is what we want.

Do you think I should keep the implementation as is? -- Or do you want me to further investigate and make the code simpler? (I am open to both)

Reply:
I think we can
1) Use the simpler implementation from Gemma
2) Add a check before the embedding layer is called in the main code to ensure that the exception is raised. We have a utility called `check_embeddings_within_bounds` in `tf_utils` for this.